rstudio: init

Mrbroggle opened the pull request:
created a module for rStudio. File is really because because all 256 terminal colours need to be maually assigned. If anyone can think of better way to do this please fix it :)

Commit:
added rstudio module

trueNAHO reviewed (changes requested):
> File is really because because all 256 terminal colours need to be maually assigned. If anyone can think of better way to do this please fix it :)

Since our colorscheme currently only provides 16 colors, the 256 colors need to converge to them. Depending on the intended behaviour, the mapping from 256 to 16 can be partially generated.

Btw, the formatting CI is failing.

Mrbroggle commented:
> Since our colorscheme currently only provides 16 colors, the 256 colors need to converge to them. Depending on the intended behaviour, the mapping from 256 to 16 can be partially generated.
> 
> Btw, the formatting CI is failing.

I don't think thats what should be done, RStudio makes you manually define each colour for the in-built terminal, so these were squashed from 256 into 16 colour you would 256 colour support for RStudio. I can do this if you think its right for the project however.

Mrbroggle committed:
Merge branch 'danth:master' into rstudio-hm-module

Commit:
fixed hm module for rstudio

Commit:
added meta.nix for rstudio

Commit:
fixed hm module naming

Commit:
removed services and added manual config push

Commit:
fixed file path for theme and made auto false

Commit:
fixed name of theme file

Commit:
fixed name of theme file again

Commit:
fixed theme colouring

Commit:
fixed theme colouring

Commit:
swapped to mustache file

Commit:
hacky work around to removed the file from the store so theme loads

Commit:
fixed formating

Commit:
fixed missing /

Commit:
fixed theme extenstion having wrong name

Commit:
added readme and added myself to maintainer

Commit:
added title to readme.md

trueNAHO commented on stylix/maintainers.nix:
Rebase this file on top of master and see how it is supposed to be used.

trueNAHO commented on modules/rstudio/meta.nix:
```diff
@@ -0,0 +1,3 @@
+{
+  maintainers = [ mrbroggle ];
```
You probably meant:

```suggestion
{ lib, ... }:
{
  maintainers = [ lib.maintainers.mrbroggle ];
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,26 @@
+            template = ./base.rstheme.mustache;
+            extension = ".rstheme.generated";
+          };
+          onChange = ''
+            rm -f ${config.xdg.configHome}/rstudio/themes/stylix.rstheme
+            cp ${config.xdg.configHome}/rstudio/themes/stylix.rstheme.generated ${config.xdg.configHome}/rstudio/themes/stylix.rstheme
+            chmod u+w ${config.xdg.configHome}/rstudio/themes/stylix.rstheme
+          '';
```
The intention behind this is to `rm --force` manual changes outside Stylix, right? I think we should remove this, following Home Manager's principle of erroring about this.

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,26 @@
+  ...
+}:
+{
+  options.stylix.targets.rstudio = {
+    enable = config.lib.stylix.mkEnableTarget "RStudio" false;
+  };
```
Consider doing the following:

```suggestion
  options.stylix.targets.rstudio.enable =
    config.lib.stylix.mkEnableTarget "RStudio" false;
  };
```

trueNAHO commented on modules/rstudio/README.md:
```diff
@@ -0,0 +1,5 @@
+
+This module is auto disabled. The theme is generated and an onChange automatically puts the file in the rstudio theme section with the correct permissions
+
+In the settings of RStudio select the theme stylixBase16 to apply the theme, any future changes will automatically change this theme.
```
Can we not somehow declare the Stylix as the default?

Mrbroggle replied:
As it stands rstudio as a package doesnt have any exposed way in nix to set a theme directly. Setting the theme once is the only way ive been able to get my module to work.

trueNAHO replied:
> I don't think that will be needed anymore, I tried again to write it with immutable files and it seems to work. Latest commit basically fixed all requested changes but 256 colours
>
> -- https://github.com/danth/stylix/pull/951/files#r2005070638

Does it now make sense to enable it by default?

trueNAHO reviewed (changes requested):
> > Since our colorscheme currently only provides 16 colors, the 256 colors need to converge to them. Depending on the intended behaviour, the mapping from 256 to 16 can be partially generated.
> >
> > [...]
>
> I don't think thats what should be done, RStudio makes you manually define each colour for the in-built terminal, so these were squashed from 256 into 16 colour you would lose 256 colour support for RStudio. I can do this if you think its right for the project however.

Any colors that are not part of the provided base16 colorscheme are hard-coded. Does the theming not look wrong when using dark and light themes?

@danth, what do you think of this?

Mrbroggle committed:
Update modules/rstudio/meta.nix

Co-authored-by: NAHO <90870942+trueNAHO@users.noreply.github.com>

Mrbroggle committed:
Update modules/rstudio/hm.nix

Co-authored-by: NAHO <90870942+trueNAHO@users.noreply.github.com>

Commit:
inital tests of changes

Mrbroggle commented:
I was going to write a whole long winded explantion of how if i wrote this module without hacky work arounds and permission changing it wouldn't work, in the process i changed my module to be more in line with the stylix ideals and now it works mostly fine. 

Without the 256 colour support, the internal terminal doesnt work looking something like this
![image](https://github.com/user-attachments/assets/046af9c1-f39d-4e31-86da-cafbfdb720bd)

Commit:
reverted to 256 colour

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
miss placed } broke terminal colours

Mrbroggle commented on modules/rstudio/README.md:
```diff
@@ -0,0 +1,5 @@
+# RStudio
+
+This module is auto disabled. The theme is generated and an onChange automatically puts the file in the rstudio theme section with the correct permissions
```
I don't think that will be needed anymore, I tried again to write it with immutable files and it seems to work.  Latest commit basically fixed all requested changes but 256 colours

trueNAHO replied:
Does this mean this part can be removed? If we also enable the module by default, we could entirely remove this `README.md` file.

awwpotato commented on modules/rstudio/meta.nix:
```diff
@@ -0,0 +1,4 @@
+{ lib, ... }:
+{
+  maintainers = [ lib.maintainers.mrbroggle ];
```
you need to add yourself to `stylix/maintainers.nix`

Mrbroggle replied:
Added myself now, lmk what else needs fixing :smile:

Mrbroggle committed:
Merge branch 'danth:master' into rstudio-hm-module

Commit:
added mrboggle to to maintainers

Mrbroggle commented:
> Does this mean this part can be removed? If we also enable the module by default, we could entirely remove this README.md file.

I think we should enable it by default if either the rStudio package or rStudioWrapper is installed, but im not sure how to do that so any pointers would be nice. Happy to change that

trueNAHO commented:
> > Does this mean this part can be removed? If we also enable the module by default, we could entirely remove this README.md file.
> 
> I think we should enable it by default if either the rStudio package or rStudioWrapper is installed, but im not sure how to do that so any pointers would be nice. Happy to change that

Usually, unless we are dealing with external applications we just enable modules by default, except if there are known problems in certain environments. For reference, here is how the NixVim module is only enabled if it has been detected:

https://github.com/danth/stylix/blob/d9df91c55643a8b5229a3ae3a496a30f14965457/modules/nixvim/nixvim.nix#L136-L140

Commit:
enabled module by default and removed useless README.md

0xda157 reviewed (commented):
you should add a testbed

danth committed:
rstudio: add testbed

danth commented:
> you should add a testbed

I added this

trueNAHO reviewed (approved):
LGTM.

0xda157 reviewed (approved):
lgtm, making a home manager module for rstudio would be a good if your up to it (as mentioned in https://github.com/danth/stylix/issues/481#issuecomment-2753863867).

danth commented:
@trueNAHO please can you check whether the conversations from [your previous review](https://github.com/danth/stylix/pull/951#pullrequestreview-2699629597) should be resolved :))

trueNAHO reviewed (changes requested):
> @trueNAHO please can you check whether the conversations from [your previous review](https://github.com/danth/stylix/pull/951#pullrequestreview-2699629597) should be resolved :))

@Mrbroggle, the dark and light themes look the same when starting the application in the testbeds:

- `nix run .#testbed:rstudio:default:dark:image:scheme:cursor`:

  ![image](https://github.com/user-attachments/assets/58e05045-6468-488d-8544-8479365597ff)

- `nix run .#testbed:rstudio:default:light:image:scheme:cursorless`:

  ![image](https://github.com/user-attachments/assets/97d8ad07-62cf-42d0-a909-6277f64e7e28)

Is it possible for the application to be themed dark depending on the theme? Maybe based on the polarity?

Mrbroggle committed:
Merge branch 'danth:master' into rstudio-hm-module

Mrbroggle commented:
> > @trueNAHO please can you check whether the conversations from [your previous review](https://github.com/danth/stylix/pull/951#pullrequestreview-2699629597) should be resolved :))
> 
> @Mrbroggle, the dark and light themes look the same when starting the application in the testbeds:
> 
> Is it possible for the application to be themed dark depending on the theme? Maybe based on the polarity?

I ran the testbeds and selected theme in rstudio and it themes correctly 
`nix run .#testbed:rstudio:default:dark:image:scheme:cursor`
![image](https://github.com/user-attachments/assets/17cdaaa3-8032-4501-a4b2-3ab10e23175a)

`nix run .#testbed:rstudio:default:light:image:scheme:cursorless`
![image](https://github.com/user-attachments/assets/d4b85b14-5d01-44e6-95f9-4cbaefd4c20f)

Did you apply the theme in settings? 
There is a file under ~/.confg/rstudio/rstudio-prefs.json that stores the editor theme but that would require changing only one line in the file and im not sure how one might do that because that file stores all the other settings for RStudio and needs RW perms which i think is out the scope of stylix?

example of rstudio-prefs.json:
```json
{
    "initial_working_directory": "~/Projects/uni/R",
    "pdf_previewer": "none",
    "posix_terminal_shell": "custom",
    "terminal_bell_style": "none",
    "highlight_r_function_calls": true,
    "editor_theme": "stylixBase16",
    "custom_shell_command": "/run/current-system/sw/bin/fish",
    "default_open_project_location": "~/Projects/uni/R",
    "relative_line_numbers": true,
    "show_diagnostics_other": true,
    "editor_keybindings": "vim",
    "text_rendering": "geometricPrecision",
    "show_rmd_render_command": true,
    "rmd_viewer_type": "pane",
    "disable_renderer_accessibility": true,
    "graphics_backend": "cairo",
    "soft_wrap_r_files": true
}
```

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

trueNAHO commented:
> > Is it possible for the application to be themed dark depending on the theme? Maybe based on the polarity?
>
> [...]
>
> Did you apply the theme in settings under `Tools > Global Options > Appearence > Editor Theme` ? There is a file under ~/.confg/rstudio/rstudio-prefs.json that stores the editor theme but that would require changing only one line in the file and im not sure how one might do that because that file stores all the other settings for RStudio and needs RW perms which i think is out the scope of stylix?

@danth, should we use a `home.activation` script?

Mrbroggle commented:
> @danth, should we use a `home.activation` script?
we could use `jq` in home.activation script something like 

```bash
mv rstudio-prefs.json temp.json
jq -r '.editor_theme|= "stylixBase16"' temp.json > rstudio-prefs.json
rm temp.json
```

danth commented:
Yes, that seems reasonable. We could use [`sponge`](https://manpages.ubuntu.com/manpages/focal/en/man1/sponge.1.html) to avoid creating a temporary file.

We'd also have to handle the case where the file hasn't been initialised yet.

Commit:
added hm activation script to enable theme

Mrbroggle commented:
all seems to be working now and auto setting the theme. It is installing jq and spone however. Which I though in the way implementend should be temporary just for this script? Anyway, let me know if any other changes need to be made.

trueNAHO committed:
rstudio: polish Bash code

trueNAHO reviewed (approved):
The theme seems to be applied now:

- `nix run .#testbed:rstudio:default:dark:image:scheme:cursor`:

  ![image](https://github.com/user-attachments/assets/e61a7851-7957-4a29-aad6-fb85aee3eca6)

- `nix run .#testbed:rstudio:default:light:image:scheme:cursorless`:

  ![image](https://github.com/user-attachments/assets/6bf10e30-99af-46dc-82c8-0e6823002bf7)

@Mrbroggle, can you approve commit c680a63090c4 ("rstudio: polish Bash code") and 545f5e37416b ("rstudio: add required meta.name attribute")?

trueNAHO committed:
rstudio: add required meta.name attribute

Add required meta.name attribute as of commit 82f67a36eba1 ("doc: align
module capitalization (#1115)").

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle commented:
> @Mrbroggle, can you approve commit [c680a63](https://github.com/danth/stylix/commit/c680a63090c4bd82e3ff363f177abd7ee6c3801f) ("rstudio: polish Bash code") and [545f5e3](https://github.com/danth/stylix/commit/545f5e37416b2d66500bc83465bff8c41eb41e17) ("rstudio: add required meta.name attribute")?

changes should be applied.

danth commented:
Can confirm, this is now working.

However, some of the text is difficult to read with a light theme:

![Screenshot of file browser](https://github.com/user-attachments/assets/7cf8aff0-49f4-481e-acbc-1eb85285cfe6)

I couldn't find which part of the template is causing this.

Mrbroggle commented:
> I couldn't find which part of the template is causing this.
I did a test, this issue is here in the mustache file:

`/* rs-theme-is-dark: TRUE */`

Im not sure how to do this programatically but we'd need to change this with the polarity.

Commit:
working theme polarity

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle commented:
I fixed it in a pretty scuffed way, i wasnt too sure how else to do it other than concat the header with the template but its working.

Commit:
fix ordering of maintainers file

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

danth commented:
> Im not sure how to do this programatically but we'd need to change this with the polarity.

All of the colors should be coming directly from the theme; the polarity option is only to control the automatically generated default theme and may not be correct otherwise. Its default is `either`.

If the RStudio theme absolutely must specify whether it is light or dark, this should be done by testing the actual colors in the theme rather than using the `polarity` value.

Commit:
not using polarity anymore, swapped to pulling base00 rgb values and calculating the luminace

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle commented:
> If the RStudio theme absolutely must specify whether it is light or dark, this should be done by testing the actual colors in the theme rather than using the `polarity` value.

To fix the difficulty to read part here

> However, some of the text is difficult to read with a light theme:
> 
> ![Screenshot of file browser](https://private-user-images.githubusercontent.com/28959268/433127373-7cf8aff0-49f4-481e-acbc-1eb85285cfe6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4NDQ1OTUsIm5iZiI6MTc0NTg0NDI5NSwicGF0aCI6Ii8yODk1OTI2OC80MzMxMjczNzMtN2NmOGFmZjAtNDlmNC00ODFlLWFjYmMtMWViODUyODVjZmU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI4VDEyNDQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0NjFlYmFhNWNkYjY2ZTBjOGM2MDlhYjQwY2Y4ZTE1MDMyZDU5ZjE3ZDVlMjNjMWFmZTNjMzllY2ZlY2NiZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fcvLB_PdrVJv2kLxVCbQv9AX-FETbrpAaKkwhOfU3Jg)

 we must set the `/* rs-theme-is-dark:  */` to either true or false so it must be set.
 
 Ive used the rgb values from base00 and calculated luminosity to set the theme-is-dark. I think this should be good.

Commit:
fixed maintainers file

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,77 @@
+                '.editor_theme |= "stylixBase16"' \
+                "$config" |
+              ${lib.getExe' pkgs.moreutils "sponge"} "$config"
+              verboseEcho \
+                "stylix: rstudio: setting editor_theme to '${name}' in $config"
```
It would be neat if this message would only occur when the previous `.editor_theme` value was not already `stylixBase16`. In that case, the following error message might be appropriate:

```suggestion

              verboseEcho \
                "stylix: rstudio: changing editor_theme from $theme to ${name} in $config"
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,77 @@
+            if [[ -f "$config" ]]; then
+              run ${lib.getExe pkgs.jq} \
+                --raw-output \
+                '.editor_theme |= "stylixBase16"' \
```
```suggestion
                '.editor_theme |= "${name}"' \
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,77 @@
+          in
+          lib.hm.dag.entryAfter [ "writeBoundary" ] ''
+            config="$HOME/.config/rstudio/rstudio-prefs.json"
+            if [[ -f "$config" ]]; then
```
Does this work in "regular" setups?


For example, running `home-manager switch` on an empty system triggers the following branch:

```nix
            else
              run cp ${file} "$config"
```

Running `home-manager switch` a second time (assuming the config file was generated) triggers the

```nix
            if [[ -f "$config" ]]; then
```

branch which attempts to override the now overwrite the read-only `/nix/store` linked file.

I assume the `[[ -f "$config" ]]` condition needs to be extended to account for this.

trueNAHO replied:
Also, it might be worth considering to add a Nix option whether overwriting the default theme to Stylix is desired. The option should default to `true`.

Mrbroggle replied:
> Does this work in "regular" setups?


Ive been running my branch with this module for the past few weeks, works fine in my setup and in testbed so it seems to work fine in a normal setup.

Mrbroggle replied:
> > Does this work in "regular" setups?
> 
> Ive been running my branch with this module for the past few weeks, works fine in my setup and in testbed so it seems to work fine in a normal setup.

Seems i got this wrong, i have sinced fixed this with a chmod to 644 perms like the file is when created by RStudio.

trueNAHO commented on modules/rstudio/testbeds/default.nix:
The application is not automatically started in the testbed. Consider using the new options from commit 8a35410a28d3 ("treewide: add stylix.testbed.ui.{command,application} options (#1110)"). You also have to use these new options as it would otherwise break when merging into master.

trueNAHO reviewed (changes requested):
> > If the RStudio theme absolutely must specify whether it is light or dark, this should be done by testing the actual colors in the theme rather than using the `polarity` value.
>
> To fix the difficulty to read part here
>
> > However, some of the text is difficult to read with a light theme:
> >
> > ![Screenshot of file browser](https://private-user-images.githubusercontent.com/28959268/433127373-7cf8aff0-49f4-481e-acbc-1eb85285cfe6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU4NDQ1OTUsIm5iZiI6MTc0NTg0NDI5NSwicGF0aCI6Ii8yODk1OTI2OC80MzMxMjczNzMtN2NmOGFmZjAtNDlmNC00ODFlLWFjYmMtMWViODUyODVjZmU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDI4VDEyNDQ1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0NjFlYmFhNWNkYjY2ZTBjOGM2MDlhYjQwY2Y4ZTE1MDMyZDU5ZjE3ZDVlMjNjMWFmZTNjMzllY2ZlY2NiZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fcvLB_PdrVJv2kLxVCbQv9AX-FETbrpAaKkwhOfU3Jg)
>
> we must set the `/* rs-theme-is-dark:  */` to either true or false so it must be set.
>
> Ive used the rgb values from base00 and calculated luminosity to set the theme-is-dark. I think this should be good.

The light theme LGTM now:

![image](https://github.com/user-attachments/assets/e893a65e-9a85-4fef-8ac0-5c941dad2327)

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
updated testbeds and applied NAHO's recomendations including options to disable file modifications

Commit:
fixed permission on created file for rstudio-prefs

Commit:
fixed missing run on chmod

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
removed from theme in verbose echo, will add later, changed how luminace values are retrived

Commit:
changed the testbed to be in line with the rest of the packages

Commit:
fixed testbed to actually run now

Mrbroggle commented:
Not sure why this CI/CD is failing, something to do with a failed build of jypter and gyp not being installed, testbed runs on my local machine though.

Commit:
rstudioWrapper seems to work with the testbed, needs testing on gh ci/cd

Commit:
testing testbed rstudio with added gyp and distutils dependecies manually added

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
undid manual package addition and added back btop testbed that was randomly removed from my repo?

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+              extension = ".rstheme";
+            };
+          };
+        home.activation.rstudioThemeSelect =
+          if config.stylix.targets.rstudio.themeAutoset then
+            let
+              file = builtins.toFile "rstudio-prefs.json" (
+                builtins.toJSON { editor_theme = name; }
+              );
+              name = "stylixBase16";
+            in
+
+            lib.hm.dag.entryAfter [ "writeBoundary" ] ''
+              config="$HOME/.config/rstudio/rstudio-prefs.json"
+              if [[ -f "$config" ]]; then
+                run ${lib.getExe pkgs.jq} \
+                  --raw-output \
+                  '.editor_theme |= "${name}"' \
+                  "$config" |
+                ${lib.getExe' pkgs.moreutils "sponge"} "$config"
+                verboseEcho \
+                  "stylix: rstudio: changing editor_theme to ${name} in $config"
+              else
+                run cp ${file} "$config"
+                run chmod 644 "$config"
+              fi
+            ''
+          else
+            lib.hm.dag.entryAfter ''
+
+              verboseEcho \
+                "stylix: rstudio: not changing editor_theme: option disabled"
+            '';
```
Does `home.activation` not support `lib.mkIf`:

```nix
home.activation.rstudioThemeSelect = let
  /* ... */
in
  lib.mkIf config.stylix.targets.rstudio.themeAutoset /* ... */;
```

Also, it would be convenient if there is no `verboseEcho` when the option is disabled.

Mrbroggle replied:
not too sure, however i couldn't get it to work

Mrbroggle replied:
> Does `home.activation` not support `lib.mkIf`:
> 
> ```nix
> home.activation.rstudioThemeSelect = let
>   /* ... */
> in
>   lib.mkIf config.stylix.targets.rstudio.themeAutoset /* ... */;
> ```
> 
> Also, it would be convenient if there is no `verboseEcho` when the option is disabled.

Ive tried with lib.mkIf as described, this does not work. However doing mkIf as `activation lib.mkIf  /*...*/ = { rstudioThemeSelect. /*...*/` does work. Although it does not look nice, it works. If someone else can get it to run with the way you sugested im all ears.

trueNAHO replied:
> Ive tried with lib.mkIf as described, this does not work. However doing mkIf as `activation lib.mkIf /*...*/ = { rstudioThemeSelect. /*...*/` does work. Although it does not look nice, it works. If someone else can get it to run with the way you sugested im all ears.

LGTM.

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+                  '.editor_theme |= "${name}"' \
+                  "$config" |
+                ${lib.getExe' pkgs.moreutils "sponge"} "$config"
+                verboseEcho \
```
Nitpick:

```suggestion

                verboseEcho \
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+                  --raw-output \
+                  '.editor_theme |= "${name}"' \
+                  "$config" |
+                ${lib.getExe' pkgs.moreutils "sponge"} "$config"
```
Is the formatter not happy with the following indentation:

```suggestion
                  ${lib.getExe' pkgs.moreutils "sponge"} "$config"
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+
+            lib.hm.dag.entryAfter [ "writeBoundary" ] ''
+              config="$HOME/.config/rstudio/rstudio-prefs.json"
+              if [[ -f "$config" ]]; then
```
Nitpick:

```suggestion

              if [[ -f "$config" ]]; then
```

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+          let
+            inherit (config.lib.stylix) colors;
+            luminance =
+              (0.2126 * (lib.toInt colors.base00-rgb-r))
+              + (0.7152 * (lib.toInt colors.base00-rgb-g))
+              + (0.0722 * (lib.toInt colors.base00-rgb-b));
```
What are these magic values?

Mrbroggle replied:
these magic values calculate luminance from RGB values from here [Relative Luminance](https://en.wikipedia.org/wiki/Relative_luminance)

awwpotato replied:
you should leave a comment or refactor it to make the code self explanatory

Mrbroggle replied:
fixed with the latest commit

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+              (0.2126 * (lib.toInt colors.base00-rgb-r))
+              + (0.7152 * (lib.toInt colors.base00-rgb-g))
+              + (0.0722 * (lib.toInt colors.base00-rgb-b));
+            polarity =
+              if luminance < 128 then
+                "/* rs-theme-is-dark: TRUE */"
+              else
+                "/* rs-theme-is-dark: FALSE */";
```
If anything, this should be checked based on `stylix.polarity`.

Mrbroggle replied:
this is from going what @danth has said 

> If the RStudio theme absolutely must specify whether it is light or dark, this should be done by testing the actual colors in the theme rather than using the polarity value.

trueNAHO reviewed (changes requested):
> Not sure why this CI/CD is failing, something to do with a failed build of jypter and gyp not being installed

Maybe the `pkgs.rstudio` derivation is large enough to timeout our CI?

Mrbroggle committed:
Apply suggestions from code review

Co-authored-by: NAHO <90870942+trueNAHO@users.noreply.github.com>

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
fixed maintainers

Commit:
added comments for various things and changed script to run on lib.mkIf

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

trueNAHO commented on modules/rstudio/hm.nix:
```diff
@@ -0,0 +1,85 @@
+  options.stylix.targets.rstudio = {
+    enable = config.lib.stylix.mkEnableTarget "RStudio" true;
+
+    themeAutoset = lib.mkOption {
+      description = ''
+        Whether to modify the RStudio preferences file to set the theme.
+      '';
+      type = lib.types.bool;
+      default = true;
+    };
```
What about:

```suggestion
    themeAutoset = lib.mkEnableOption "setting Stylix as the theme" // {
      default = true;
      example = false;
    };
```

Mrbroggle committed:
Apply suggestions from code review

Co-authored-by: NAHO <90870942+trueNAHO@users.noreply.github.com>

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Mrbroggle commented on modules/rstudio/base.rstheme.mustache:
appologies for the taking a while to get back to this MR. As it stands i dont see a way to either dynamically genrerate the xterm colours even similar to the blender module. Like for some stupid reason the theme needs all those colours or the terminal just turns black and loses all colour. 

If there is a way to generate the colours id love to use it to reduce the file by about 2k line. [This](https://www.ditig.com/256-colors-cheat-sheet#xterm-non-system-colors) the exact list of colours needed maybe the module could fetch these from another source but even then i dont think thats the best course of action.

There is also the option to make a dictionary of the xterm colours similar to [this vim plugin ](https://github.com/guns/xterm-color-table.vim/blob/master/plugin/xterm-color-table.vim) and use that to generate the rest with nix and append it to the bottom of the file. 

As it stands these colours need to be set to xterm colours, its just about reducing file size imo

Mrbroggle committed:
Merge branch 'master' into rstudio-hm-module

Commit:
changed to import a list then create tags to reduce the size of the files and make it a little easier to make changes later to dynamically generate the xterm colours

trueNAHO reviewed (changes requested):
The https://github.com/nix-community/stylix/pull/951#discussion_r2107648663 discussion seems unaddressed.

Could you also sync again with master?
